File: My-DSA-Portfolio/Sorting Algos/MergeSort/MergeSort.java

import java.util.ArrayList;
public class MergeSort {
    void merge(int[] arr, int low, int mid, int high){
        int left = low;
        int right = mid + 1;
        ArrayList<Integer> temp = new ArrayList<>();
        // merging two sorted arrays
        while( left <= mid && right <= high){
            if(arr[left] <= arr[right]){
                temp.add(arr[left]);
                left ++;
            }
            else{
                temp.add(arr[right]);
                right ++;
            }
        }
        // left part
        while( left <= mid){
            temp.add(arr[left]);
            left ++;
        }
        // right part
        while( right <= high){
            temp.add(arr[right]);
            right ++;
        }
        // copy the elements from temp to original array
        for(int i = 0 ; i < temp.size(); i++){
            arr[low + i] = temp.get(i);
        }
    }
    void mergeSort(int[] arr, int low, int high){
        if( low >= high){
            return;
        }
        int mid = low + (high - low ) / 2;
        mergeSort(arr, low, mid);
        mergeSort(arr, mid + 1, high);
        merge(arr, low, mid, high);
    }
}

File: 1. Sorting Algos/4. Merge Sort/merge_sort.py
class MergeSort:
    def merge(self, arr, low, mid, high):
        left = low
        right = mid + 1
        temp = []
        
        # Merge the two halves into temp
        while left <= mid and right <= high:
            if arr[left] <= arr[right]:
                temp.append(arr[left])
                left += 1
            else:
                temp.append(arr[right])
                right += 1
        
        # Copy any remaining elements of the left half
        while left <= mid:
            temp.append(arr[left])
            left += 1
        
        # Copy any remaining elements of the right half
        while right <= high:
            temp.append(arr[right])
            right += 1
        
        # Copy the sorted elements back into the original array
        for i in range(len(temp)):
            arr[low + i] = temp[i]

    def mergeSort(self, arr, low, high):
        # Base case: if the array has one or no elements, it's already sorted
        if low >= high:
            return
        
        mid = low + (high- low) // 2
        
        # Recursively sort the two halves
        self.mergeSort(arr, low, mid)
        self.mergeSort(arr, mid + 1, high)
        
        # Merge the sorted halves
        self.merge(arr, low, mid, high)

    def display(self, arr):
        print(arr)

if __name__ == "__main__":
    arr = [1, 8, 9, 3, 7, 5, 2, 2, 1]
    m = MergeSort()
    m.mergeSort(arr, 0, len(arr) - 1)
    m.display(arr)